subprojects/integ-test/src/integTest/groovy/org/gradle/integtests/SyncTaskIntegrationTest.groovy

```diff
@@ -394,36 +393,71 @@ class SyncTaskIntegrationTest extends AbstractIntegrationSpec {
-    def "change in case of input file still syncs properly"() {
+    def "change in case of input files and folders will still syncs properly"() {
         given:
-        buildFile << """
-            task writeFile {
-                def outputFile = new File(buildDir, project.hasProperty("capitalize") ? "OUTPUT" : "output")
-                outputs.file outputFile
-                doLast {
-                    outputFile.text = "hello"
-                }
-            }
-            task sync(type: Sync) {
-                from writeFile
-                into new File(buildDir, "sync")
+        buildFile << '''
+            task syncIt(type: Sync) {
+                from project.hasProperty("capitalizeFile") ? "FILE.TXT" : "file.txt"
+                from project.hasProperty("capitalizeDir") ? "DIR" : "dir"
+                into buildDir
             }
-        """
+        '''
+
+        def lowercaseFile = file('file.txt')
+        lowercaseFile.text = 'file'
+
+        file('dir').create {
+            file('file2.txt') << 'file2'
+            file('file3.txt') << 'file3'
+            file('file4.txt') << 'file4'
+        }
+
+        and:
+        run 'syncIt'
+        file('build').assertHasDescendants(
+            'file.txt',
+            'file2.txt',
+            'file3.txt',
+            'file4.txt'
+        )
+
+        and:
+        lowercaseFile.delete() // renameTo wouldn't change the case
+        def uppercaseFile = file('FILE.TXT')
+        uppercaseFile.text = 'FILE'
+        assert uppercaseFile.canonicalFile.name == 'FILE.TXT'
+        and:
+        file('dir/file3.txt').delete()
+
+        file('dir/file4.txt').delete()
+        file('dir/File4.txt') << 'File4'
+
         when:
-        succeeds("sync")
+        succeeds('syncIt', '-PcapitalizeFile')
+        def uppercaseOutputFile = file('build/FILE.TXT')
         then:
-        file("build/output").assertExists()
-        file("build/sync/output").assertExists()
+        executedAndNotSkipped ':syncIt'
+        uppercaseOutputFile.parentFile.listFiles() != [].toArray()
+        uppercaseOutputFile.assertExists()
+        uppercaseOutputFile.text == 'FILE'
+        uppercaseOutputFile.canonicalFile.name == 'FILE.TXT'
+        and:
+        file('build/file2.txt').assertExists()
+        file('build/file3.txt').assertDoesNotExist()
+        file('build/File4.txt').assertExists()
+        file('build/File4.txt').canonicalFile.name == 'File4.txt'
 
         when:
-        succeeds("sync", "-Pcapitalize")
+        file('dir').deleteDir()
+        file('DIR').create {
+            file('file4.txt') << 'file4'
+        }
+        and:
+        succeeds('sync', '-PcapitalizeDir')
         then:
-        file("build/OUTPUT").assertExists()
-        // TODO: This should exist
-        file("build/sync/OUTPUT").assertDoesNotExist()
+        executedAndNotSkipped ':syncIt'
+        file('build/file4.txt').canonicalFile.name == 'file4.txt'
```

Comment:
There is some test coverage missing when the case of a directory changes. For that you would need to create a sub-directory in `DIR` and then change its case. I guess the task will fail if the directory contains some files.

What should we do in this case?

Reply:
I am changing the case of `dir` to `DIR` (couldn't use rename, had to delete and recreate to make sure the case is changed correctly) in https://github.com/gradle/gradle/pull/9663/files/3bba35426f9a9faec675b3b00e22883d531030fb#diff-98f85d28fc88be35783c5f9ab2192e19R453

Reply:
Yes, you are changing the case of `dir` to `DIR`, though you don't sync the directory, you only sync files in the directory. That means that no dir with a case change is in the output directory.

Reply:
Understood, added a case to handle it.
Also changed fs delete to fs rename, seems less invasive.

Reply:
Validated on Mac, Windows and Linux